### ISSUE TITLE
File vs File descriptor

### ISSUE THREAD

**Issue by schomatis** [ipfs/go-mfs#8]
The `fileDescriptor` structure seems to be doing too much work which should be moved to the `File` structure.

**Comment by schomatis** [comment]
The `fileDescriptor` seems to be basically a wrapper for the `DagModifier` but the most confusing part is that it keeps a reference to the `File` that created it (`inode`) accessing all of its attributes, making these two very coupled structures.

**Comment by schomatis** [comment]
Could we collapse both under `File`? (Moving the `DagModifier` from `fileDescriptor`.) It already seems like a file can only have one file descriptor operating on it. So instead of keeping a reference to the file descriptor we would only operate through `File`.

@Stebalien WDYT?

**Comment by schomatis** [comment]
> It already seems like a file can only have one file descriptor operating on it.

Taking a closer look at the `desclock` it seems that it's design is meant to allow either one R/W file descriptor open or *many* RO file descriptors open simultaneously.

https://github.com/ipfs/go-mfs/blob/166e30ed671e5dd067643b646969bf1deb51cdbd/file.go#L75-L79

**Comment by Stebalien** [comment]
I believe the idea is that a File represents the file itself and a file descriptor is a handle to an *open* file. Basically, having a File is like calling `open("/path...", 0)`. From there, one can move/rename the file but one can't read/write it without opening it.

As you noted, it *also* allows for many readers. However, that can actually be implemented without ever *exposing* the underlying "file" object to the user (that can be an implementation detail).

**Comment by schomatis** [comment]
> review the API boundaries between the two and better define who does what.

One thing to review is the relationship between `DagModifier` in the file descriptor and the `ipld.Node` in the file (from which the first one is created) and clearly define which one reflects the current state of the file and during  which moments during MFS operaions.

**Comment by nitishm** [comment]
> > review the API boundaries between the two and better define who does what.
> 
> One thing to review is the relationship between `DagModifier` in the file descriptor and the `ipld.Node` in the file (from which the first one is created) and clearly define which one reflects the current state of the file and during which moments during MFS operaions.

So the `DagModifier` has a `dagserv` which is a reference to the `File(.inode).dagService`. Yet in a most places we use the `DagModifier` while in some we use `fd.inode.dagservice` 
https://github.com/ipfs/go-mfs/blob/e5a375ded4ae3dfbf6fca4eb1c5e606f19fdc676/fd.go#L135

Why not just use the `fd.mod.dagserv` instead ? This is really confusing (too many `dagservice`s!

**Comment by schomatis** [comment]
> This is really confusing (too many `dagservice`s!

I can definitely sympathize with that, in almost all circumstances they are all accessing the same DAG service (created in the `setupNode` and related methods) so you can just think of it as *the* DAG service. Why everyone has one? Depends on the case, I too would like to reduce a bit the number of references to it, but sometimes it's simpler to copy the reference than to access someone else's.

**Comment by aschmahmann** [comment]
This repository is no longer maintained and has been copied over to [Boxo](https://github.com/ipfs/boxo).  In an effort to avoid noise and crippling in the Boxo repo from the weight of issues of the past, we are closing most issues and PRs in this repo.  Please feel free to open a new issue in Boxo (and reference this issue) if resolving this issue is still critical for unblocking or improving your usecase.

You can learn more in the [FAQs for the Boxo repo copying/consolidation effort](https://github.com/ipfs/boxo/wiki/Copied-or-Migrated-Repos-FAQ).
